### PR TITLE
Refactor clientApi some

### PR DESCRIPTION
**Pull request by Bjvanminnen**
I'm starting to look at the client API in preparation for sharing it better. This set of changes isn't strictly necessary, but I think makes things a little cleaner. Ultimately, I expect this code will end up moving into shared and also being used by initApp.

What I changed:
- Fewer nested classes. Instead we have a top level class (ClientApiRequest) and a number of others that inherit it. I then moved methods on the child classes into the parent. This means some of these may have methods they don't need, but this still feels cleaner. We could explicitly disallow those methods for subclasses if we were concerned.
- Changed names. It felt to me like it was easy to think SharedTable was an object instead of an API that provides us methods to manipulate data over the wire.

Let me know if you disagree with these changes, or see anything I missed.

### CODE REVIEW

File: apps/src/clientApi.js [hunk 1]
```diff
@@ -142,11 +161,21 @@ ApiRequestHelper.prototype.delete = function (localUrl, callback) {
...
+ * @param {!string} id
+ * @param {!Object} value
+ * @param {!NodeStyleCallback} callback
+ */
+ClientApiRequest.prototype.update = function (id, value, callback) {
+  this.post('/' + id, value, callback);
+};
+
+/**
+ * @param {!string|number} id Id to delete
```
Comment (islemaster) [inline apps/src/clientApi.js]: The `!` in your JSDoc type here defines `id` as a required parameter (i.e. something you should never pass `undefined` to) but below you have a special case for handling `id === undefined` that appears valid.  Maybe this should actually be an optional parameter?
```
 * @param {string|number} [id] Id to delete
```

File: apps/src/clientApi.js [hunk 2]
```diff
@@ -123,12 +134,20 @@ ApiRequestHelper.prototype.postToGet = function (localUrl, data, callback) {
...
 /**
+ * @param {!Object} value
+ * @param {!NodeStyleCallback} callback
+ */
+ClientApiRequest.prototype.create = function (value, callback) {
+  this.postToGet('', value, callback);
+};
+
+/**
  * @param {!string} localUrl - API endpoint relative to API base URL.
```
Comment (islemaster) [inline apps/src/clientApi.js]: JSDoc needs to be updated to match changed parameter name below.

File: apps/src/clientApi.js (hunk 1 above)
Comment (Bjvanminnen) [inline apps/src/clientApi.js]: Ahh yes, for some reason I was thinking ! meant optional at the time.

I wouldn't think we would ever actually call this without an id, but the way the code existed before it was possible to do so (on the helper class). I wonder if the better fix would be to have it be required, and remove the undefined check.

**Comment by islemaster** [comment]
When I originally implemented this API wrapper I had an inheritance relationship something like what you've done here. @davidsbailey pushed back against it and argument for composition over inheritance in this case, citing the "refused bequest" code smell and how the property bag should have all the behaviors that the base class implements. [See the original discussion here](https://github.com/code-dot-org/code-dot-org/pull/1207#discussion_r24281167) that led to the current form.

File: apps/src/clientApi.js (hunk 1 above)
Comment (islemaster) [inline apps/src/clientApi.js]: Oh yeah, if we never use it that way let's make it required.

**Comment by Bjvanminnen** [comment]
@davidsbailey Would you like to weigh in here? I know when we were talking yesterday, you said that the inheritance patterns in this file ended up feeling a little bit awkward to you. Do you think this new approach makes things better/worse? Do you think other things could be done to make it even better?

File: apps/src/clientApi.js (hunk 1 above)
Comment (Bjvanminnen) [inline apps/src/clientApi.js]: I don't know for sure that we don't, but I suspect we don't. Making it required is slightly riskier, but is cleaner if accurate. I think I'll make the change, and watch out for things breaking.

File: apps/src/clientApi.js [hunk 3]
```diff
@@ -49,12 +50,7 @@ require('./utils');
  *        method being invoked.
  */
 
-/** 
- * Namespace for the client API for accessing channels, tables and properties.
- */
-var clientApi = module.exports;
-
-var ApiRequestHelper = function (baseUrl) {
+var ClientApiRequest = function (baseUrl) {
```
Comment (islemaster) [inline apps/src/clientApi.js]: Tiny nit: I never had JSDoc on `ApiRequestHelper` here (I should have) but removal of the doc'd `var clientApi = ...` line has caused my linter to incorrectly associate the JSDoc callback definition above with this line.  Might be nice to doc this.
```
/**
 * Base class for API request helpers.
 * @param {string} baseUrl
 * @constructor
 */
var ClientApiRequest = function (baseUrl) {
```

**Comment by islemaster** [comment]
It looks like you didn't change the interface at all so it's a non-issue, but for reference here's [the fake version of this in netsimTestUtils.js](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/test/util/netsimTestUtils.js#L31). It should still represent the API correctly (it implements `readAll`, `read`, `create`, `update`, `delete`, and its own test-specific method `log`).

**Comment by islemaster** [comment]
Here's (maybe) the weirdest example:

The `set` and `update` methods on `ClientApiRequest` now run exactly the same code.  Other than names, they are identical.  However, they have a different semantic - you shouldn't pass any old string key to the `update` method, because they underlying API you expect to interact with will choke on different things.

Effectively we have a D.R.Y. bottleneck:
```
Different callers
     \    /
  Identical API helper code
     /    \
Different APIs
```

In this case, the same set of operations is being applied in two different circumstances.  We do this all the time, but when we do it's usually by pulling in a "helper" - a utility or external library - and not expressed through inheritance.  I think that's what originally led Dave and I to think inheritance was the wrong type of code reuse for this situation.  The decision happens to be really fuzzy in this case because the different circumstances are very similar, but we thought not quite similar enough.

**Comment by davidsbailey** [comment]
LGTM. This fixes my main gripe with what's in staging, namely that SharedTable and and UserTable inherit from Channel, which is not very OO because it is not true conceptually that "a SharedTable is a Channel" or "a UserTable is a Channel". 

Brad, in the discussion of your previous PR, I believe I sited the Refused Bequest smell when I saw that the child class was redefining or ignoring inherited methods in a way that rejected the behavior "Bequested" by the parent. Extracting a helper may have eliminated this code smell but did not eliminate the conceptually inappropriate inheritance relationship.

Hope this clears things up. Thanks for fixing, Brent!

**Comment by Bjvanminnen** [comment]
Regarding set vs. update, having set seems a little strange to me in general. I would expect to have a create and an update. Is set meant to essentially be "create if not exists, else update"?

**Comment by islemaster** [comment]
I approve this change, and the discussion surrounding it. :dizzy: The more you know!

**Comment by davidsbailey** [comment]
A potentially related problem with the underlying Properties API is that it uses a bare JSON value (string, number, boolean) as the value. REST APIs are supposed to use fully-formed JSON objects (which includes only Objects or Arrays).

**Pull request merged**